src/App.tsx

```diff
@@ -1,25 +1,62 @@
-import React from 'react';
+import { useEffect, useMemo, useState } from 'react';
+import { useLocation, useNavigate } from 'react-router-dom';
 import './App.css';
 import { getNumbers } from './utils';
+import { Pagination } from './components/Pagination';
 
 // eslint-disable-next-line @typescript-eslint/no-unused-vars
-const items = getNumbers(1, 42).map(n => `Item ${n}`);
 
 export const App: React.FC = () => {
+  const location = useLocation();
+  const navigate = useNavigate();
+
+  const queryParams = new URLSearchParams(location.search);
+  const initialPage = parseInt(queryParams.get('page') || '1', 10);
+  const initialPerPage = parseInt(queryParams.get('perPage') || '5', 10);
+
+  const [currentPage, setCurrentPage] = useState(initialPage);
+  const [perPage, setPerPage] = useState(initialPerPage);
+
+  const items = useMemo(() => getNumbers(1, 42).map(n => `Item ${n}`), []);
+
+  useEffect(() => {
+    const params = new URLSearchParams();
+
+    params.set('page', currentPage.toString());
+    params.set('perPage', perPage.toString());
+    navigate({ search: `?${params.toString()}` });
+  }, [currentPage, perPage, navigate]);
+
+  const startIndex = (currentPage - 1) * perPage;
+  const currentItems = items.slice(startIndex, startIndex + perPage);
+
+  const handlePageChange = (page: number) => {
+    setCurrentPage(page);
+  };
+
+  const handlePerPageChange = (event: React.ChangeEvent<HTMLSelectElement>) => {
+    setPerPage(parseInt(event.target.value, 10));
+    setCurrentPage(1);
+  };
+
   return (
     <div className="container">
       <h1>Items with Pagination</h1>
 
       <p className="lead" data-cy="info">
-        Page 1 (items 1 - 5 of 42)
+        Page {currentPage} (items {startIndex + 1} -{' '}
+        {Math.min(startIndex + perPage, items.length)} of {items.length})
```

Comment:
save result in a variable above and just use here instead of doing this computation is JSX